transaction/core/src/ring_signature/rct_bulletproofs.rs

```diff
@@ -155,11 +213,32 @@ impl SignatureRctBulletproofs {
             decompressed_pseudo_output_commitments.push(commitment);
         }
 
-        // Compute the pedersen generators for this token_id
-        let generator = generators(token_id);
+        // Collect list of of unique token ids
+        let token_ids = {
+            let mut token_ids = BTreeSet::default();
+            token_ids.insert(fee.token_id);
+            for token_id in &self.output_token_ids {
+                token_ids.insert(token_id.into());
+            }
+            for token_id in &self.pseudo_output_token_ids {
+                token_ids.insert(token_id.into());
+            }
+            token_ids
+        };
+
+        // Get a generator cache
+        let mut generator_cache = GeneratorCache::default();
```

Comment:
1. Yeah, in the current design for MCIP #25, we are going with generator per token id, and so this object is caching those.
2. Yeah, the generators are computed here: https://github.com/mobilecoinfoundation/mobilecoin/blob/93647188e543f5a9eaaf6e2a49587ebfd828a512/transaction/core/src/ring_signature/mod.rs#L46 This formula has to be fixed pretty much for all time or older tokens will not be recognized. The formula is a bit funky, mainly because, the way we computed the initial generator for MOB was a bit funky, and we have to be backwards compatible to that forever.
3. The idea here was that this cache is built for every transaction. The thing is, the cache as it currently exists is not constant time. The idea I had here was like, we're not trying to avoid revealing how many token ids you have in your Tx (which would be basically infeasible because we have different range proofs per token id), or, which TxOut's have the same token id as which other TxOut's or pseudo outputs. (We could maybe do that). We don't want to reveal when a token id used by Bob is the same as the token id used by Alice. So, we trash the cache when we look at a new transaction, to try to kill off cache timing attacks.

This means that e.g. if your transaction has 5 inputs and outputs, and two token ids, you do two elliptic curve operations to get generators, instead of 5 before the cache.

We could make this better from a privacy point of view if we had a constant-time map for ristretto points, but currently that is impossible because ristretto point doesn't implement `subtle::ConditionalAssign`. But also the constant-time map we have in the enclave now is pretty dumb and just does linear scanning, so if that thing can live forever and get very big i might get worried about performance. But it wouldn't get big unless there are a lot of token ids. But someday that may happen.

I feel like doing a lot of work to avoid 2 elliptic curve operations when validating a Tx may be misguided -- I was able to eliminate two elliptic curve operations anyways during validation by doing `RistrettoPoint::identity()` instead of `generator.commit(Zero, Zero)`. I think there is a lot of low hanging optimization fruit right now. So I feel like engineering a massive, fast, constant-time global generator cache might not be the easiest way to make it faster.

We might be able to do something like, hard-code a table of the first 16 token ids or something, and constant-time scan across that, then fall back to the btree map? That might be worth doing?

Reply:
Dumping the cache makes a lot sense. It might make sense for us to take a quick look at how much work putting `subtle::ConditionalAssign` into Dalek Ristretto might be. If it's low hanging fruit it might worth it! I can take a peek at what that would entail. 

I agree with you that there's potentially other easier optimizations to do in the meantime - although at this very moment, I'm not sure what they are. But I don't even think they're needed for this PR. It's already big. But subsequent 'quickening' PRs might be a good idea.

I guess importantly, we're not currently taking a measure of the speed. It sounds like a wise thing to do to potentially put benchmarks gauge this?

Reply:
I think we used to have some benchmarks targets, but they weren't in CI and eventually got stale and then got deleted

I remember joey used to run benchmarks of transaction construction when we were deciding how big rings should be

We should make some more cargo benchmarks. It would be great if we actually tracked those numbers over time, we could have something that watches for perf regressions when a PR goes in

Reply:
> f it's low hanging fruit it might worth it! I can take a peek at what that would entail.

:+1: 

> potentially other easier optimizations to do in the meantime - although at this very moment, I'm not sure what they are.

I tried to make some tickets.. let me see if i can track them down...

I opened this PR a while ago: https://github.com/mobilecoinfoundation/mobilecoin/pull/1423/files

I think this "if check_value_is_preserved" is a client-side operation that is done when building a transaction, to confirm that consensus will agree that the tx is valid, before we submit it. But the way it is done today, it entails a bunch of elliptic curve operations on the phone. I think we might be able to do it better if we just check that that, sum of values (u64's) is balanced, and sum of blindings (dalek::Scalar) is balanced.

I suspect that the "scalar mul precomputation table" that Isis talked about (I made a ticket here: https://github.com/mobilecoinfoundation/mobilecoin/issues/1564) may have significantly more impact, but it's also going to entail either:
1. Sticking that table in a lazy static, which would work for existing targets (and which we currently do for bulletproofs generators), but is possibly a tech debt
2. Passing the scalar mul table around everywhere (a lot of work, but more flexibility for embedded devices)